src/test/java/integration/EventsTest.java

```diff
@@ -778,6 +780,122 @@ public void given_multibranchWithSources_when_createEventForExistingBranch_then_
     }
 
     @Test
+    @Issue("JENKINS-42511")
+    public void given_multibranchWithSources_when_createEventWhileIndexing_then_onlyOneBuildCreated() throws Exception {
+        try (MockSCMController c = MockSCMController.create()) {
+            c.createRepository("foo");
+            BasicMultiBranchProject prj = r.jenkins.createProject(BasicMultiBranchProject.class, "foo");
+            prj.setCriteria(null);
+            prj.getSourcesList().add(new BranchSource(new MockSCMSource(null, c, "foo", true, false, false)));
+
+            // Now we need some latency, this will ensure that the indexing and the event are interleaved during
+            // processing
+            c.withLatency(MockLatency.fixed(100, TimeUnit.MILLISECONDS));
+            // we have not received any events so far, so in fact this should be 0L
+            long lastModified = prj.getComputation().getEventsFile().lastModified();
+
+            QueueTaskFuture<Queue.Executable> future = prj.scheduleBuild2(0).getFuture();
+            SCMHeadEvent.fireNow(new MockSCMHeadEvent(SCMEvent.Type.CREATED, c, "foo", "master", c.getRevision("foo", "master")));
+            future.get();
+            FreeStyleProject master = prj.getItem("master");
+            waitUntilNoActivityIgnoringThreadDeathUpTo(10000);
+
+            assertThat(prj.getComputation().getEventsFile().lastModified(), greaterThan(lastModified));
+            assertThat("The master branch was built once only", master.getLastBuild().getNumber(), is(1));
+            assertThat(master.getLastBuild().getResult(), is(Result.SUCCESS));
+            List<Throwable> deaths = new ArrayList<Throwable>();
+            for (Computer comp : r.jenkins.getComputers()) {
+                for (Executor e : comp.getExecutors()) {
+                    if (e.getCauseOfDeath() != null) {
```

Comment:
I would suggest just taking #92 first, which is simple and I think uncontroversial on its own (would defend against all manner of other mistakes), and then adapting this to verify that the revised lesser symptoms do not occur—namely, a failed indexing run.